README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->
distcrete
=========

[![Project Status: Active – The project has reached a stable, usable state and is being actively developed.](http://www.repostatus.org/badges/latest/active.svg)](http://www.repostatus.org/#active)

[![Travis-CI Build Status](https://travis-ci.org/reconhub/distcrete.svg?branch=master)](https://travis-ci.org/reconhub/distcrete) [![Build status](https://ci.appveyor.com/api/projects/status/uy5out71p9qkh93i/branch/master?svg=true)](https://ci.appveyor.com/project/thibautjombart/distcrete/branch/master) [![codecov.io](https://codecov.io/github/reconhub/distcrete/coverage.svg?branch=master)](https://codecov.io/github/reconhub/distcrete?branch=master)

> Discrete Distribution Approximations

`distcrete` takes a distribution and a set of parameters and returns a list object with functions as elements. Each one is the equivalent to the function calls we typically expect to be able to do in R for a given distribution:

-   Density e.g.`dnorm`
-   Distribution function e.g. `pnorm`
-   Quantile function e.g. `qnorm`
-   Random generation e.g. `rnorm`

Each function created by `distcrete` corresponds to the first letter of the non-discrete equivalent.

``` r
set.seed(415)
d0 <- distcrete::distcrete("gamma", 1, shape = 3, w = 0)
d0$d(1:10)
#>  [1] 0.243022187 0.253486335 0.185086776 0.113451286 0.062683215
#>  [6] 0.032332641 0.015882196 0.007521773 0.003462799 0.001558522
d0$p(c(.1,.5))
#> [1] 0.09958372 0.19115317
d0$q(c(.1,.5))
#> [1] 0 1
d0$r(10)
#>  [1]  2  3  1  2 12  2  2  4  5  3
```

Allowed distributions
---------------------

You can use any distribution that conforms to the following expectations:

-   It has a distribution function like `p[dist name]` available
-   It has a quantile function like `q[dist name]` available

These can be loaded from a package or created on the fly, but must exist when the `distcrete()` function is called.

Installation
------------

You can install it from CRAN with:

``` r
install.packages("distcrete")
```

You can install `distcrete` the most up to date version from github with:

``` r
# install.packages("devtools")
devtools::install_github("reconhub/distcrete")
```

Tests
-----

``` r
devtools::test()
#> Loading distcrete
#> Loading required package: testthat
#> Testing distcrete
#> distcrete: .....................................................................................................................................
#> utils: ...............
#> 
#> DONE ======================================================================
```


NEWS.md
# distcrete 1.0.3
* New to CRAN





tests/testthat.R
library(testthat)
library(distcrete)

test_check("distcrete")


tests/testthat/test-utils.R
context("utils")

test_that("assert_in_range", {
  expect_null(assert_in_range(0.5, 0, 1))
  expect_error(assert_in_range(1.1, 0, 1),
               "must be in range")
  expect_null(assert_in_range(1, 0, 1),
               "must be in range")
})

test_that("assert scalar", {
  expect_error(assert_scalar(NULL), "must be a scalar")
  expect_error(assert_scalar(1:2), "must be a scalar")
})

test_that("assert_numeric", {
  expect_error(assert_numeric(NULL), "must be numeric")
  expect_error(assert_numeric("a"), "must be numeric")
  expect_error(assert_numeric(TRUE), "must be numeric")

  expect_null(assert_numeric(1L))
  expect_null(assert_numeric(NA_real_))
})

test_that("assert_nonmissing", {
  expect_error(assert_nonmissing(NA, "must not be NA"))
  expect_error(assert_nonmissing(c(1, NA), "must not be NA"))
})

test_that("assert_function", {
  expect_error(assert_function(NULL, "must be a function"))
  expect_null(assert_function(sin))
  expect_null(assert_function(assert_function))
})


tests/testthat/test-distcrete.R
context("distcrete")

EPS <- 1e-12

test_that("basic", {
  shape <- 3
  rate <- 1.15

  for (w in c(0, 0.25, 0.5, 0.99)) {
    d <- distcrete("gamma", 1, shape, rate, w = w)
    x <- seq(0, 20, by = d$interval)

    expect_equal(sum(d$d(x)), d$cdf(max(x) + 1 - w), tolerance = EPS)
    px <- d$p(x)
    if (w == 0) {
      expect_equal(px, d$cdf(x + 1), tolerance = EPS)
    }

    expect_equal(d$q(px), x, tolerance = EPS)
    expect_equal(d$p(d$q(px)), px, tolerance = EPS)
  }
})

## This is the simplest case; the distribution is defined over the
## infinite interval so there is basically no information loss as we
## move among the different representations of data.
test_that("infinite domain distribution", {
  dx <- 0.1
  x1 <- seq(-5, 5, by = dx) # Covering almost the whole range
  x2 <- seq(-1, 2, by = dx) # Covering just the high density region

  for (w in c(0, 1/3, 0.5, 1)) {
    d <- distcrete("norm", dx, w = w)

    ## Check that we're targeting the correct distribution
    expect_equal(d$cdf(x1), pnorm(x1), tolerance = EPS)

    p1 <- d$d(x1)
    p2 <- d$d(x2)
    ## This is the total range that we cover:
    r1 <- c(min(x1) - dx * w, max(x1) + dx * (1 - w))
    r2 <- c(min(x2) - dx * w, max(x2) + dx * (1 - w))
    expect_equal(sum(p1), diff(d$cdf(r1)), tolerance = EPS)
    expect_equal(sum(p2), diff(d$cdf(r2)), tolerance = EPS)

    ## p and q functions can roundtrip appropriately
    px1 <- d$p(x1)
    expect_equal(d$q(px1), x1, tolerance = EPS)
    expect_equal(d$p(d$q(px1)), px1, tolerance = EPS)

    px2 <- d$p(x2)
    expect_equal(d$q(px2), x2, tolerance = EPS)
    expect_equal(d$p(d$q(px2)), px2, tolerance = EPS)

    expect_equal(d$cdf(r1[[1]]) + cumsum(p1),
                 d$cdf(x1 + dx * (1 - w)), tolerance = EPS)
    expect_equal(d$cdf(r2[[1]]) + cumsum(p2),
                 d$cdf(x2 + dx * (1 - w)), tolerance = EPS)
  }
})

## Next up, let's get a semi-infinite case tested.  This is a bit
## nastier because we need to deal with information loss on the left
## boundary.
test_that("semi-infinite domain distribution", {
  dx <- 0.1
  x1 <- seq(0, 10, by = dx) # Covering almost the whole range
  x2 <- seq(0, 4, by = dx) # Covering just the high density region
  shape <- 3
  rate <- 1.15

  for (w in c(0, 1/3, 0.5, 1)) {
    d <- distcrete("gamma", dx, shape, rate, w = w)

    ## Check that we're targeting the correct distribution:
    expect_equal(d$cdf(x1), pgamma(x1, shape, rate), tolerance = EPS)

    p1 <- d$d(x1)
    p2 <- d$d(x2)

    ## This is the total range that we cover:
    r1 <- c(min(x1) - dx * w, max(x1) + dx * (1 - w))
    r2 <- c(min(x2) - dx * w, max(x2) + dx * (1 - w))
    expect_equal(sum(p1), diff(d$cdf(r1)), tolerance = EPS)
    expect_equal(sum(p2), diff(d$cdf(r2)), tolerance = EPS)

    ## p and q functions can roundtrip appropriately
    px1 <- d$p(x1)
    expect_equal(d$q(px1), x1, tolerance = EPS)
    expect_equal(d$p(d$q(px1)), px1, tolerance = EPS)

    px2 <- d$p(x2)
    expect_equal(d$q(px2), x2, tolerance = EPS)
    expect_equal(d$p(d$q(px2)), px2, tolerance = EPS)

    expect_equal(d$cdf(r1[[1]]) + cumsum(p1),
                 d$cdf(x1 + dx * (1 - w)), tolerance = EPS)
    expect_equal(d$cdf(r2[[1]]) + cumsum(p2),
                 d$cdf(x2 + dx * (1 - w)), tolerance = EPS)
  }
})

## With a uniform distribution it should be pretty easy to work with
## this, except that the _pair_ boundaries make it very lossy
## unfortunately.
test_that("boundary conditions", {
  dx <- 0.1
  x <- seq(0, 1, by = dx)
  n <- length(x)

  for (w in c(0, 1/3, 0.5, 1)) {
    d <- distcrete("unif", dx, w = w)

    ## Check that we're targeting the correct distribution
    expect_equal(d$cdf(x), punif(x), tolerance = EPS)

    p <- d$d(x)
    r <- c(min(x) - dx * w, max(x) + dx * (1 - w))
    expect_equal(sum(p), diff(d$cdf(r)), tolerance = EPS)

    ## p and q functions can roundtrip appropriately
    px <- d$p(x)
    qpx <- d$q(px)
    if (w < 1) {
      expect_equal(qpx, c(x[-n], x[n - 1]), tolerance = EPS)
    } else {
      expect_equal(qpx, x)
    }

    if (w > 0 && w < 1) {
      expect_equal(d$p(qpx), c(px[-n], px[n - 1]), tolerance = EPS)
    } else {
      expect_equal(d$p(qpx), px, tolerance = EPS)
    }

    expect_equal(d$cdf(r[[1]]) + cumsum(p),
                 d$cdf(x + dx * (1 - w)), tolerance = EPS)
  }
})

test_that("out of bounds", {
  dx <- 0.1
  x <- seq(0, 1, by = dx)
  w <- 0
  d <- distcrete("unif", dx, w = w)

  ## The CDF is defined over the whole interval
  expect_equal(d$p(-1), 0) # same as punif(-1)
  expect_equal(d$p(2), 1) # same as punif(2)

  ## The PDF is also defined over this interval
  expect_equal(d$d(-1), 0) # same as dunif(-1)
  expect_equal(d$d(2), 0) # same as dunif(2)

  ## The quantile function is not though
  expect_equal(suppressWarnings(d$q(-1)), NaN) # same as qunif(-1)
  expect_equal(suppressWarnings(d$q(2)), NaN) # same as dunif(2)
})

test_that("strict mode", {
  dx <- 0.1
  x <- seq(0, 1, by = dx)
  w <- 0
  d <- distcrete("unif", dx, w = w)

  expect_error(d$p(dx / 2, strict = TRUE),
               "Values do not align")
  expect_error(d$d(dx / 2, strict = TRUE),
               "Values do not align")
})

test_that("random number generation", {
  dx <- 0.1
  x <- seq(0, 1, by = dx)
  for (w in c(0, 1/3, 0.5, 1)) {
    d <- distcrete("unif", dx, w = w)

    p <- runif(10000)
    table(d$q(p))
    table(d$q(p) + d$interval)

    z <- d$r(1000)
    expect_true(all(z %in% x))
    r <- c(if (w < 1) 0.0 else 0.1,
           if (w > 0) 1.0 else 0.9)
    expect_equal(range(z), r)
  }
})

test_that("logspace", {
  shape <- 3
  rate <- 1.15
  d <- distcrete("gamma", 1, shape, rate, w = 0)
  x <- seq(0, 20, by = d$interval)

  expect_equal(d$d(x, log = TRUE), log(d$d(x)))

  ## Step it out to bigger values and see if we can help:
  xx <- c(20, 30, 40, 50)

  p <- d$d(xx, log = TRUE)
  expect_true(all(is.finite(p)))
  expect_true(all(p < dgamma(xx, shape, rate, log = TRUE)))

  p2 <- d$d(xx)
  i <- p2 == 0
  expect_true(any(i))
  expect_equal(exp(p[!i]), p2[!i])
})

test_that("print method, no args", {
  d <- distcrete("norm", 1)
  capture.output(expect_identical(d, print(d)))
  str <- capture.output(print(d))
  expect_match(str, "name: norm", all = FALSE)
  expect_match(str, "with no parameters", all = FALSE)
})

test_that("print method, withargs", {
  d <- distcrete("norm", 1, mean = 1)
  capture.output(expect_identical(d, print(d)))
  str <- capture.output(print(d))
  expect_match(str, "name: norm", all = FALSE)
  expect_match(str, "mean: 1", all = FALSE)

  ## Unnamed:
  expect_equal(capture.output(print(distcrete("norm", 1, 1))), str)
})


R/util.R
assert_in_range <- function(x, min, max, name = deparse(substitute(x))) {
  assert_scalar_numeric(x, name)
  if (x < min || x > max) {
    stop(sprintf("%s must be in range [%s, %s]", name, min, max))
  }
}

assert_scalar <- function(x, name=deparse(substitute(x))) {
  if (length(x) != 1) {
    stop(sprintf("%s must be a scalar", name), call.=FALSE)
  }
}

assert_numeric <- function(x, name=deparse(substitute(x))) {
  if (!is.numeric(x)) {
    stop(sprintf("%s must be numeric", name), call.=FALSE)
  }
}

assert_nonmissing <- function(x, name=deparse(substitute(x))) {
  if (any(is.na(x))) {
    stop(sprintf("%s must not be NA", name), call.=FALSE)
  }
}

assert_function <- function(x, name=deparse(substitute(x))) {
  if (!is.function(x)) {
    stop(sprintf("%s must be a function", name), call.=FALSE)
  }
}

assert_scalar_numeric <- function(x, name = deparse(substitute(x))) {
  assert_scalar(x, name)
  assert_numeric(x, name)
  assert_nonmissing(x, name)
}


R/distcrete.R
#' Discretise a distribution.
#' @title Discretise a distribution
#' @param name The name of a distribution function (e.g.,
#'   \code{norm}, \code{gamma}).  The distribution must have a cdf
#'   function (e.g., \code{pnorm}) and a quantile function (e.g.,
#'   \code{qnorm}) defined.
#'
#' @param interval The interval to discretise the interval onto.
#'
#' @param ... Parameters to \code{cdf}.  Can be matched positionally
#'   or by name.
#'
#' @param w How to weight the endpoints; must be between 0 and 1.  
#'   If 0.5 then integration happens centred around the interval, if
#'   0 floor, if 1 then ceiling.
#'
#' @param anchor Any location that is a valid \code{x}
#' @export
#' @author Rich FitzJohn
#' @examples 
#' library(distcrete)
#' set.seed(415)
#' d0 <- distcrete("gamma", 1, shape = 3, w = 0)
#' d0$d(1:10)
#' d0$p(c(.1,.5))
#' d0$q(c(.1,.5))
#' d0$r(10)



distcrete <- function(name, interval, ..., w = 0.5, anchor = 0) {
  ## TODO: the offset / reverse parts should be done against the
  ## returned objects because they alter only the range?

  ## TODO: What about non-uniform specification of interval?

  ## NOTE: Once we have Anne's thing in here it's mostly going to be a
  ## case of storing a series of the integrations of u f(u) du for a
  ## range of values; we can immediately do that over basically the
  ## whole range and store it as a vector (do this for all numbers
  ## from (eps to 1-eps) and do the rest on demand).  Alternatively we
  ## could probably just compute it against a spline.

  cdf <- match.fun(paste0("p", name))
  qf <- match.fun(paste0("q", name))
  assert_in_range(w, 0, 1)
  assert_scalar_numeric(anchor)
  d <- list(cdf = function(x, log = FALSE) cdf(x, ..., log.p = log),
            qf = function(x, log = FALSE) qf(x, ..., log.p = log),
            interval = interval,
            w = w,
            anchor = anchor,
            ## offset = 0,
            ## reverse = FALSE,
            parameters = list(...))
  d$d <- function(x, log = FALSE, strict = FALSE) distcrete_d(d, x, log, strict)
  d$p <- function(q, log = FALSE, strict = FALSE) distcrete_p(d, q, log, strict)
  d$q <- function(p, log = FALSE) distcrete_q(d, p, log)
  d$r <- function(n = 1) distcrete_r(d, n)
  d$name <- name
  class(d) <- "distcrete"
  d
}

## The nice thing about this is that it holds for _all_ offsets.  But
## I think that we do want to get this for a given anchor.
distcrete_d <- function(d, x, log = FALSE, strict = FALSE) {
  ## TODO: what is the logic here?  This seems like a 1st order
  ## appoximation perhaps?  Worth documenting this properly somewhere
  check_interval(x, d$anchor, d$interval, d$w, strict)
  x0 <- x - d$w * d$interval
  p0 <- d$cdf(x0, log)
  p1 <- d$cdf(x0 + d$interval, log)
  if (log) {
    log_minus(p1, p0)
  } else {
    p1 - p0
  }
}

## TODO: make the argument log.p rather than log?
distcrete_p <- function(d, q, log = FALSE, strict = FALSE) {
  check_interval(q, d$anchor, d$interval, d$w, strict)
  d$cdf(q + (1 - d$w) * d$interval, log)
}

## TODO: I'm still not 100% sure about the -1 here!
distcrete_q <- function(d, p, log = FALSE) {
  x <- d$qf(p, log)
  find_interval(x, d$anchor, d$interval, d$w, FALSE) - d$interval
}

## TODO: consider a lower/uppper argument to _q so that this
## (+interval-interval) bit can be avoided...
distcrete_r <- function(d, n = 1, ...) {
  distcrete_q(d, stats::runif(n)) + d$interval
}

##' @export
print.distcrete <- function(x, ...) {
  cat("A discrete distribution\n")
  cat(sprintf("  name: %s\n", x$name))
  p <- x$parameters
  ## This might collect a few too many parameters, but will at least
  ## capture defaults.
  if (is.null(names(p)) || !all(nzchar(names(p)))) {
    f <- environment(x$cdf)$cdf
    p <- as.list(match.call(f, as.call(c(list(quote(f)), list(1), p))))[-(1:2)]
  }
  if (length(p) == 0L) {
    cat("  <with no parameters>\n")
  } else {
    cat(sprintf("  parameters:\n"))
    to_str <- function(x) if (length(x) == 1) as.character(x) else ""
    cat(sprintf("    %s: %s\n", names(p), vapply(p, to_str, character(1))),
        sep="")
  }
  invisible(x)
}

## TODO: find a reference for this bit of hackery:
log_minus <- function(lx, ly) {
  dyx <- ly - lx
  lx + ifelse(dyx > -log(2), log(-expm1(dyx)), log1p(-exp(dyx)))
}

## This is going to be the workhorse for re-relating continuous 'x'
## onto our new scale.
find_interval <- function(x, anchor, interval, w, strict = FALSE) {
  ## (fabs((x) - R_forceint(x)) > 1e-7*fmax2(1., fabs(x)))
  xs <- x / interval - anchor + w
  ## "safely" drop all the floating point noise
  xs <- zapsmall(xs, -trunc(log10(.Machine$double.eps) / 2))
  if (strict) {
    if (any(xs != floor(xs))) {
      stop("Values do not align")
    }
  } else {
    xs <- floor(xs)
  }
  xs * interval + anchor
}

check_interval <- function(x, anchor, interval, w, strict = FALSE) {
  if (strict) {
    find_interval(x, anchor, interval, w, strict)
  }
}


vignettes/src/ml.R
## ---
## title: "ML estimation with distcrete"
## author: "Thibaut Jombart"
## date: "`r Sys.Date()`"
## output: rmarkdown::html_vignette
## vignette: >
##   %\VignetteIndexEntry{distcrete_ML}
##   %\VignetteEngine{knitr::rmarkdown}
##   %\VignetteEncoding{UTF-8}
## ---

##+ echo = FALSE, results = "hide"
knitr::opts_chunk$set(
  error = FALSE,
  fig.width = 7,
  fig.height = 5)
set.seed(1)


## In this vignette, we illustrate how *distcrete* can be combined with
## classical optimisation procedures available in R to derive maximum-likelihood
## (ML) of parameters of distributions of interest. We simulate simple data from
## a discretised exponential distribution, and then attempt to recover the
## original parameters using ML.


## ## Simulating data

## Simulating data from a discretised distribution is very easy using
## *distcrete*, as discretised distributions already contain a random variate
## generator in the `$r` component. We illustrate this by generating a
## discretised exponential distribution `x` with rate parameter 0.123:
library(distcrete)
rate <- 0.123
plot(function(x) dexp(x, rate), xlim = c(0, 60),
     main = "Original Exponential distribution")
x <- distcrete("exp", interval = 1L, rate)
x$r


## We simulate 200 draws from the distribution `x`:
##+ simulation
set.seed(1)
sim <- x$r(200)
head(sim)
summary(sim)
plot(table(sim), lwd=3, xlab = "x", ylab = "Frequency",
     main = "Simulated sample")




## ## ML estimation

## We will use the base function `optim` to find maximum likelihood estimates of
## the rate of the distribution `x`. The log-likelihood function to be optimised
## can be written as:

## + target
ll <- function(param) {
    d <- distcrete("exp", interval = 1L, param)$d
    sum(d(sim, log = TRUE))
}


## For instance, the log-likelihood with shapes and rates of 1.234 can be computed as:
##+ ll_example
param <- c(rate = 1.234)
ll(param)



## We can feed this function and initial parameters to `optimise`:
##+ optimise
opt <- optimise(ll, c(0, 20), maximum = TRUE)
opt


## The ML estimate for the rate of the distribution is `r opt$maximum`, which is
## close enough to the original value of `r rate`.




## ## Discretised gamma example

## We repeat the same exercise with a gamma distribution. The only difference
## with the previous example is that `optim` needs to be used, as the parameter
## space has now 2 dimensions.

## We start by simulating a sample from a discretised gamma distribution:
##+ gamma
shape <- 10
rate <- 1.2
y <- distcrete("gamma", interval = 1L, shape, rate)
set.seed(1)
sim2 <- y$r(300)

head(sim2)
summary(sim2)
plot(table(sim2), lwd = 10, xlab = "x", ylab = "Frequency",
     main = "Simulated sample")



## We create a log-likelihood function for these data, as well as a deviance (it
## is easier to minimise the deviance with `optim`):
##+ ll_gamma
ll2 <- function(param) {
   d <- distcrete("gamma", interval = 1L, param[1], param[2])$d
   sum(d(sim2, log = TRUE))
}

dev2 <- function(param) -2 * ll2(param)

optim(c(1,1), dev2)


## We can verify that the estimates are not dependent on the initial state:
##+ verif
optim(c(0.5,20), dev2)


## Finally, we can plot the likelihood surface to see how tightly the distribution fits the data:

x <- seq(0, 20, length = 30)
y <- seq(0, 5, length = 30)
grid <- expand.grid(x, y)
names(grid) <- c("shape", "rate")
ll2.val <- apply(grid, 1, ll2)
df <- cbind.data.frame(grid, ll = ll2.val)

library(ggplot2)
ggplot(df, aes(x = shape, y = rate, fill = ll)) +
    geom_raster() +
    geom_contour(aes(z = ll), color = "black")

## Interestingly, it seems that quite a few combinations of parameters of the
## gamma distribution can lead to similar fits. Let us examine a few:

param1 <- c(5, 0.5)
param2 <- c(20, 2)

d1 <- distcrete("gamma", interval = 1L, param1[1], param1[2])
d2 <- distcrete("gamma", interval = 1L, param2[1], param2[2])

plot(table(d1$r(300)), lwd = 10, xlab = "x", ylab = "Frequency",
           main = "shape: 5, rate: 0.1")

plot(table(d2$r(300)), lwd = 10, xlab = "x", ylab = "Frequency",
           main = "shape: 20, rate: 2")

